### PR TITLE
Make it compatible with IMDSv2

### PR DESCRIPTION
**Pull request by daiyyr**
## Description

Make it compatible with IMDSv2

## Motivation

Fix https://github.com/aws/aws-toolkit-azure-devops/issues/531

## Related Issue(s), If Filed

https://github.com/aws/aws-toolkit-azure-devops/issues/531

## Checklist

-   [x] I have read the **README** document
-   [x] I have read the **CONTRIBUTING** document
-   [x] My code follows the code style of this project
-   [x] I have added tests to cover my changes
-   [x] A short description of the change has been added to the changelog using the script `npm run newChange`

## License

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by rknechtel** [comment]
**Please** merge this PR and release a **new version** of the AWS Toolkit. The current version is almost 2 years old. And I made the mistake of updating my EC2 instances to IMDSv2 and now all my S3 copies to my EC2 instances out of Azure DevOps are failing!

**Comment by rknechtel** [comment]
Attn: Maintainers

When will this PR be merged in? I'm running into an issue that I am sure has to do with the lack of IMDSv2 support in the AWS Toolkit. I updated all my EC2 instances to IMDSv2 and suddenly all my S3 copies from S3 to my EC2 instances (using SSM) in my Azure DevOps pipelines stopped working. I contacted AWS Support but they have been no help. I asked them about when a new version of the AWS Toolkit would be released as it's been almost 2 years since the last release. I mentioned this specific PR to them.

It appears to me that AWS and the maintainers have abandoned this project.

**Review by rbbarad — APPROVED** [review]
@daiyyr @rknechtel Thanks for your contribution! I've added some changes to this pr to make it compatible with both IMDSv1 and v2 instance types. Please take a look and reach out if there are any concerns!

File: src/lib/awsConnectionParameters.ts [hunk 1]
```diff
@@ -282,26 +282,78 @@ export async function getCredentials(awsParams: AWSConnectionParameters): Promis
...
+        token = ''
+    }
+
+    return (await getRegionFromImds(metadataService, token as string)) as string
+}
+
+/**
+ * Attempts to get a Instance Metadata Service V2 token
+ */
+async function getImdsV2Token(metadataService: AWS.MetadataService) {
```
Comment (bryceitoc9) [inline src/lib/awsConnectionParameters.ts:304]:
```suggestion
async function getImdsV2Token(metadataService: AWS.MetadataService): Promise<string> {
```

File: src/lib/awsConnectionParameters.ts [hunk 2]
```diff
@@ -282,26 +282,78 @@ export async function getCredentials(awsParams: AWSConnectionParameters): Promis
...
                 }
+            }
+        )
+    })
+}
+
+/**
+ * Attempts to get the region from the Instance Metadata Service
+ */
+async function getRegionFromImds(metadataService: AWS.MetadataService, token: string) {
```
Comment (bryceitoc9) [inline src/lib/awsConnectionParameters.ts:330]: This eliminates the trailing `as string` on line 298

```suggestion
async function getRegionFromImds(metadataService: AWS.MetadataService, token: string): Promise<string> {
```

File: src/lib/awsConnectionParameters.ts [hunk 3]
```diff
@@ -282,26 +282,78 @@ export async function getCredentials(awsParams: AWSConnectionParameters): Promis
...
-    return new Promise<string>((resolve, reject) => {
-        const metadataService = new AWS.MetadataService()
-        metadataService.httpOptions.timeout = 5000
-        metadataService.request('/latest/dynamic/instance-identity/document', (err, data) => {
-            try {
+    const imdsOptions = { httpOptions: { timeout: 5000 }, maxRetries: 2 }
+
+    const metadataService = new AWS.MetadataService(imdsOptions)
+
+    let token
```
Comment (bryceitoc9) [inline src/lib/awsConnectionParameters.ts:289]: This eliminates the `token as string` on line 298

```suggestion
    let token: string
```

File: src/lib/awsConnectionParameters.ts [hunk 4]
```diff
@@ -282,26 +282,78 @@ export async function getCredentials(awsParams: AWSConnectionParameters): Promis
...
+
+    let options = {}
 
-                console.log('...received instance identity document from metadata')
-                const identity = JSON.parse(data)
-                if (identity.region) {
-                    resolve(identity.region)
+    if (token) {
+        options = { headers: { 'x-aws-ec2-metadata-token': token } }
+    }
```
Comment (bryceitoc9) [inline src/lib/awsConnectionParameters.ts:337]: * prefer consts over lets if possible
* prefer being explicit about strings being empty strings if there's no possibility it's undefined.
```
const options = token !== '' ? { headers: { 'x-aws-ec2-metadata-token': token } } : {}
```

**Review by bryceitoc9 — APPROVED** [review]
Mostly stylistic nits: won't block merge.

**Comment by daiyyr** [comment]
Thank you @rknechtel and @rbbarad !!

**Pull request merged**